src/preload/preload_interface.h

```diff
@@ -434,6 +435,24 @@ struct preload_thread_locals {
   PTR(struct msghdr) notify_control_msg;
 };
 
+// The set of flags that can be set for each fd in syscallbuf_fds_disabled.
+enum syscallbuf_fd_flags_mask {
+  // Absent any other flags specifying special handling, this operation should
+  // always be traced.
+  FD_FLAG_TRACE_ALL = 0x1,
+  // pwrites to this fd may proceed untraced during record, but
+  // require an assist during replay
+  FD_FLAG_PWRITE_REPLAY_ASSIST = 0x3,
+  // Writes to this FD will not block
+  // N.B.: At the moment this shares a flag value with
+  // FD_FLAG_WRITE_REPLAY_ASSIST. This may be separated in the future.
+  FD_FLAG_WRITE_WONT_BLOCK = 0x2,
```

Comment:
Why only replay-assist pwrite? Why not all writes handled by `on_syscall_exit`?

Also, it's not clear to me why `FD_FLAG_PWRITE_REPLAY_ASSIST` shares bits with `FD_FLAG_TRACE_ALL` and how this would be extended. If I wanted to make proc-mem reads not blockable and not tracing then it's not clear to me how to do that.

Also I don't think it's safer to or-together a FD_FLAG_WRITE_WONT_BLOCK bit for one file descriptor with a zero flag for another file descriptor. As you've written it, bits need to add more conservative behavior.

Possibly the way to go here is, instead of using bits, define a set of fd classes "untraced", "traced", "proc-mem (reads untraced and non-blocking, writes replay-assist and non-blocking)" that we can extend later, and when combining the values instead of or-ing bits, if two classes are different, collapse to "traced". What do you think?

Reply:
In that case we'd continue to use helper functions like `fd_writes_block`, which would implement the behavior for each class.

Reply:
> Why only replay-assist pwrite?

For other writes, at least when the fd is /proc/pid/mem, we need to record the file descriptor offset during record. If we replay assist that, we wouldn't see it during record and wouldn't be able to write down the offset.

I agree that the flags are a bit cumbersome. When I originally wrote this, I had assumed that the flags would need to be known to replay, so I wanted to build in some flexibility for unknown flags and make it backwards compatible with just having the one flag that disables buffering from old traces.

However, I since realized that the flags don't actually matter during replay at all, since we record any modification we make to them on the rr side. As a result, there's no backwards compatibility issues with completely changing the scheme.

Given that, I think switching it to an enum makes sense. I'll make the appropriate modification.

Reply:
Actually, I just realized one issue with having it be categories (that I knew about when I wrote this, but had forgotten again): Right now, we have one set of these flags per address space, but as we repeatedly find out that doesn't necessarily mean that every task in the address space has the same file descriptor table. I guess that makes the current patch incorrect as well...

Reply:
Right, but I suggested

> when combining the values instead of or-ing bits, if two classes are different, collapse to "traced".

Doesn't that address the problem?

Reply:
Not necessarily, since it's possible for a different task to have an fd with the same number that was syscallbuf-opened, but the tracer doesn't know about. I guess we could also have, for each fd a count of how many times it was opened by the syscallbuf without telling the tracer about it it. Then, if a class is set and that number is non-zero, we consider that to mean "traced" also.

Reply:
> a different task to have an fd with the same number that was syscallbuf-opened, but the tracer doesn't know about.

So you're saying that in some case(s) it's a problem to fall back to the traced-syscall path for I/O on an fd that was opened using syscallbuf-open? What case would that be? That sounds like it would be a problem for all sorts of reasons.

Reply:
No, that's fine the problem problem is something like:
Thread 1 opens fd 5 through the syscall buf without setting it in the tracer
Thread 2 opens /proc/pid/mem as fd 5, the tracer sets the fd kind to proc_mem
Thread 1 reads from fd 5 and gets the wrong answer for fd_writes_block (it also does an unnecessary replay assist - but that we could probably filter during replay).

Reply:
I see. So the problem is that the "/proc/.../mem" enum value is actually *less* conservative from the default "untraced" state.

Reply:
Can we just disallow these less-conservative states when different tasks in the address space don't all have the same fd table?

Reply:
> Can we just disallow these less-conservative states when different tasks in the address space don't all have the same fd table?

I think that might be ok.